File: app/src/main/java/com/commonpepper/photosen/ui/activities/SinglePhotoActivity.java
package com.commonpepper.photosen.ui.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;

import com.commonpepper.photosen.Photosen;
import com.commonpepper.photosen.R;
import com.commonpepper.photosen.network.DownloadService;
import com.commonpepper.photosen.network.model.Photo;
import com.commonpepper.photosen.network.model.PhotoSizes;
import com.commonpepper.photosen.ui.fragments.CommentsFragment;
import com.commonpepper.photosen.ui.fragments.PhotoDetailsFragment;
import com.getkeepsafe.taptargetview.TapTarget;
import com.getkeepsafe.taptargetview.TapTargetSequence;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.squareup.picasso.Picasso;

import java.util.List;

public class SinglePhotoActivity extends AbstractNavActivity {
    private static final String TAG = SinglePhotoActivity.class.getSimpleName();

    public static final String PHOTO_TAG = "PARCELABLE_PHOTO";
    public static final String PHOTOS_URL = "https://www.flickr.com/photos/";

    private static final String PREFERENCE_FIRST_LAUNCH = "SINGLE_PHOTO_FIRST_LAUNCH";

    private Photo photo;

    private DownloadService.Aciton action;
    private PhotoDetailsFragment detailsFragment;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_single_photo);

        Toolbar toolbar = findViewById(R.id.single_image_toolbar);
        drawerLayout = findViewById(R.id.drawer_layout);
        NavigationView navigationView = findViewById(R.id.nav_view);

        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        navigationView.setNavigationItemSelectedListener(this);

        photo = getIntent().getParcelableExtra(PHOTO_TAG);

        TextView title = findViewById(R.id.single_photo_title);
        ImageView imageView = findViewById(R.id.single_photo_image_view);

        FloatingActionButton fabDownload = findViewById(R.id.fab_download);
        FloatingActionButton fabSetWallpaper = findViewById(R.id.fab_wallpaper);
        hideFAB(fabDownload);
        hideFAB(fabSetWallpaper);
        fabDownload.setOnClickListener(v -> {
            action = DownloadService.Aciton.DOWNLOAD_ONLY;
            if (Photosen.isStoragePermissionGranted(SinglePhotoActivity.this)) {
                downloadPhoto();
            }
        });
        fabSetWallpaper.setOnClickListener(v -> {
            action = DownloadService.Aciton.WALLPAPER;
            if (Photosen.isStoragePermissionGranted(SinglePhotoActivity.this)) {
                downloadPhoto();
            }
        });

        if (photo.getTitle() != null && photo.getTitle().length() > 0) {
            title.setText(photo.getTitle());
        } else {
            title.setVisibility(View.GONE);
        }

        Picasso.get().load(photo.getUrl_z()).into(imageView);

//        OLD VERSION:
//        if (getSupportFragmentManager().getFragments().size() == 0) {
//            detailsFragment = PhotoDetailsFragment.newInstance(photo);
//            getSupportFragmentManager().beginTransaction()
//                    .add(R.id.single_image_details_layout, detailsFragment)
//                    .commit();
//        } else {
//            detailsFragment = (PhotoDetailsFragment) getSupportFragmentManager().getFragments().get(0);
//        }

//        LET'S GET FRAGMENT LINKS IN CASE OF SCREEN ROTATION
        CommentsFragment commentsFragment;
        boolean detailsFlag = false;
        boolean commentsFlag = false;
        for (Fragment fragment : getSupportFragmentManager().getFragments()) {
            if (fragment instanceof PhotoDetailsFragment) {
                detailsFragment = (PhotoDetailsFragment) fragment;
                detailsFlag = true;
            }
            if (fragment instanceof CommentsFragment) {
                commentsFragment = (CommentsFragment) fragment;
                commentsFlag = true;
            }
        }
        if (!detailsFlag) {
            detailsFragment = PhotoDetailsFragment.newInstance(photo);
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.single_image_details_layout, detailsFragment)
                    .commit();
        }
        if (!commentsFlag) {
            commentsFragment = CommentsFragment.newInstance(photo.getId());
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.single_image_comments_layout, commentsFragment)
                    .commit();
        }
//        END OF RESTORING FRAGMENT LINKS

        detailsFragment.getLiveSizes().observe(this, x -> {
            showFAB(fabDownload);
            showFAB(fabSetWallpaper);

            SharedPreferences prefs = getSharedPreferences(Photosen.PREFERENCES, MODE_PRIVATE);
            boolean firstLaunch = prefs.getBoolean(PREFERENCE_FIRST_LAUNCH, true);
            if (firstLaunch) {
                new TapTargetSequence(this).targets(
                        TapTarget.forView(fabDownload, getString(R.string.download_fab_hint))
                                .tintTarget(false),
                        TapTarget.forView(fabSetWallpaper, getString(R.string.set_as_wallpaper_fab_hint))
                                .tintTarget(false)
                ).continueOnCancel(true).start();
                prefs.edit().putBoolean(PREFERENCE_FIRST_LAUNCH, false).apply();
            }

            imageView.setOnClickListener(v -> {
                List<PhotoSizes.SizesBean.SizeBean> sizes = x.getSizes().getSize();
                String url = sizes.get(sizes.size() - 1).getSource();
                Intent intent = new Intent(SinglePhotoActivity.this, PreviewActivity.class);
                intent.putExtra(PreviewActivity.TAG_URL, url);
                startActivity(intent);
            });
        });
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    private void openInBrowser(String url) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        startActivity(intent);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.single_image_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.single_image_open_in_browser) {
            openInBrowser(PHOTOS_URL + photo.getOwner() + "/" + photo.getId());
        } else if (id == R.id.single_image_share) {
            shareUrl(PHOTOS_URL + photo.getOwner() + "/" + photo.getId());
        }

        return super.onOptionsItemSelected(item);
    }

    private void downloadPhoto() {
        if (!DownloadService.isRunning) {
            Toast.makeText(this, getString(R.string.download_started), Toast.LENGTH_SHORT).show();
            Intent intent = new Intent(this, DownloadService.class);
            if (detailsFragment.getLiveSizes().getValue() != null) { //shouldn't be null, when fabs are visible, but check just in case
                List<PhotoSizes.SizesBean.SizeBean> sizes = detailsFragment.getLiveSizes().getValue().getSizes().getSize();
                String url = sizes.get(sizes.size() - 1).getSource();
                intent.putExtra(DownloadService.TAG_URL, url);
                String[] urlSplit = url.split("\\.");
                String format = urlSplit[urlSplit.length - 1];
                if (!format.equals("jpg") && !format.equals("gif") && !format.equals("png"))
                    format = "jpg";
                intent.putExtra(DownloadService.TAG_FILENAME, photo.getId() + "." + format);
                intent.putExtra(DownloadService.TAG_ACTION, action);
                startService(intent);
            }
        } else {
            Toast.makeText(this, getString(R.string.please_wait_for_current_download), Toast.LENGTH_SHORT).show();
        }
    }

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            Log.v(TAG, "Permission: " + permissions[0] + "was " + grantResults[0]);
            if (action != null) downloadPhoto();
        } else {
            Toast.makeText(this, getString(R.string.no_permisson), Toast.LENGTH_SHORT).show();
        }
    }

    private void shareUrl(String url) {
        if (photo != null) {
            Intent share = new Intent(Intent.ACTION_SEND);
            share.setType("text/plain");
            share.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);

            share.putExtra(Intent.EXTRA_SUBJECT, R.string.flickr_image);
            share.putExtra(Intent.EXTRA_TEXT, url);

            startActivity(Intent.createChooser(share, getString(R.string.share_with)));
        }
    }

    private void hideFAB(FloatingActionButton fab) {
        CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) fab.getLayoutParams();
        FloatingActionButton.Behavior behavior = (FloatingActionButton.Behavior) params.getBehavior();
        if (behavior != null) behavior.setAutoHideEnabled(false);
        fab.hide();
    }

    private void showFAB(FloatingActionButton fab) {
        fab.show();
        CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) fab.getLayoutParams();
        FloatingActionButton.Behavior behavior = (FloatingActionButton.Behavior) params.getBehavior();
        if (behavior != null) behavior.setAutoHideEnabled(true);
    }
}

File: app/src/main/java/com/commonpepper/photosen/ui/activities/SinglePhotoActivity.kt
package com.commonpepper.photosen.ui.activities

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout.LayoutParams
import androidx.core.widget.NestedScrollView
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.Observer
import com.commonpepper.photosen.Photosen
import com.commonpepper.photosen.Photosen.Companion.instance
import com.commonpepper.photosen.Photosen.Companion.isStoragePermissionGranted
import com.commonpepper.photosen.R
import com.commonpepper.photosen.R.*
import com.commonpepper.photosen.model.Photo
import com.commonpepper.photosen.model.PhotoSizes
import com.commonpepper.photosen.network.DownloadService
import com.commonpepper.photosen.network.DownloadService.Aciton
import com.commonpepper.photosen.ui.fragments.CommentsFragment
import com.commonpepper.photosen.ui.fragments.CommentsFragment.Companion.newInstance
import com.commonpepper.photosen.ui.fragments.PhotoDetailsFragment
import com.commonpepper.photosen.ui.fragments.PhotoDetailsFragment.Companion.newInstance
import com.getkeepsafe.taptargetview.TapTarget
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.floatingactionbutton.FloatingActionButton.Behavior
import com.google.android.material.navigation.NavigationView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_single_photo.*
import java.util.concurrent.Executors

class SinglePhotoActivity : AbstractNavActivity() {
    override val abstractDrawerLayout: DrawerLayout get() = drawerLayout
    private lateinit var photo: Photo
    private lateinit var detailsFragment: PhotoDetailsFragment
    private var action: Aciton? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_single_photo)
        val toolbar: Toolbar = findViewById(id.single_image_toolbar)
        val navigationView: NavigationView = findViewById(id.navigationView)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        navigationView.setNavigationItemSelectedListener(this)
        photo = intent.getParcelableExtra(PHOTO_TAG)!!
        val saveHistory = intent.getBooleanExtra(SAVE_HISTORY_TAG, true)
        val historyDao = instance.database.historyDao
        if (saveHistory) {
            Executors.newSingleThreadExecutor().execute {
                photo.time = System.currentTimeMillis()
                historyDao.insert(photo)
            }
        }
        val title: TextView = findViewById(id.single_photo_title)
        val imageView: ImageView = findViewById(id.single_photo_image_view)
        val nestedScrollView: NestedScrollView = findViewById(id.single_image_nested_sc)
        val fabDownload: FloatingActionButton = findViewById(id.fab_download)
        val fabSetWallpaper: FloatingActionButton = findViewById(id.fab_wallpaper)
        hideFAB(fabDownload)
        hideFAB(fabSetWallpaper)
        fabDownload.setOnClickListener {
            action = Aciton.DOWNLOAD_ONLY
            if (isStoragePermissionGranted(this@SinglePhotoActivity)) {
                downloadPhoto()
            }
        }
        fabSetWallpaper.setOnClickListener {
            action = Aciton.WALLPAPER
            if (isStoragePermissionGranted(this@SinglePhotoActivity)) {
                downloadPhoto()
            }
        }
        if (photo.title != null && photo.title!!.isNotEmpty()) {
            title.text = photo.title
        } else {
            title.visibility = View.GONE
        }
        Picasso.get().load(photo.url_z).into(imageView)

//        LET'S GET FRAGMENT LINKS IN CASE OF SCREEN ROTATION
        var commentsFragment: CommentsFragment? = null
        var detailsFlag = false
        var commentsFlag = false
        for (fragment in supportFragmentManager.fragments) {
            if (fragment is PhotoDetailsFragment) {
                detailsFragment = fragment
                detailsFlag = true
            }
            if (fragment is CommentsFragment) {
                commentsFragment = fragment
                commentsFlag = true
            }
        }
        if (!detailsFlag) {
            detailsFragment = newInstance(photo)
            supportFragmentManager.beginTransaction()
                    .add(id.single_image_details_layout, detailsFragment)
                    .commit()
        }
        if (!commentsFlag) {
            commentsFragment = newInstance(photo.id)
            supportFragmentManager.beginTransaction()
                    .add(id.single_image_comments_layout, commentsFragment)
                    .commit()
        }
//        END OF RESTORING FRAGMENT LINKS


        commentsFragment!!.setNestedScrollView(nestedScrollView)
        detailsFragment.liveSizes.observe(this, Observer { x: PhotoSizes ->
            showFAB(fabDownload)
            showFAB(fabSetWallpaper)
            val prefs: SharedPreferences = getSharedPreferences(Photosen.PREFERENCES, Context.MODE_PRIVATE)
            val firstLaunch = prefs.getBoolean(PREFERENCE_FIRST_LAUNCH, true)
            if (firstLaunch) {
                TapTargetSequence(this).targets(
                        TapTarget.forView(fabDownload, getString(string.download_fab_hint))
                                .tintTarget(false),
                        TapTarget.forView(fabSetWallpaper, getString(string.set_as_wallpaper_fab_hint))
                                .tintTarget(false)
                ).continueOnCancel(true).start()
                prefs.edit().putBoolean(PREFERENCE_FIRST_LAUNCH, false).apply()
            }
            imageView.setOnClickListener {
                val sizes = x.sizes!!.size
                val url = sizes!![sizes.size - 1].source
                val intent = Intent(this@SinglePhotoActivity, PreviewActivity::class.java)
                intent.putExtra(PreviewActivity.TAG_URL, url)
                startActivity(intent)
            }
        })
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    private fun openInBrowser(url: String) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.single_image_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.single_image_open_in_browser) {
            openInBrowser(PHOTOS_URL + photo.owner + "/" + photo.id)
        } else if (id == R.id.single_image_share) {
            shareUrl(PHOTOS_URL + photo.owner + "/" + photo.id)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun downloadPhoto() {
        if (!DownloadService.isRunning) {
            Toast.makeText(this, getString(string.download_started), Toast.LENGTH_SHORT).show()
            val intent = Intent(this, DownloadService::class.java)
            if (detailsFragment.liveSizes.value != null) { //shouldn't be null, when fabs are visible, but check just in case

                val sizes = detailsFragment.liveSizes.value!!.sizes!!.size
                val url = sizes!![sizes.size - 1].source
                intent.putExtra(DownloadService.TAG_URL, url)
                val urlSplit = url!!.split("\\.").toTypedArray()
                var format = urlSplit[urlSplit.size - 1]
                if (format != "jpg" && format != "gif" && format != "png") format = "jpg"
                intent.putExtra(DownloadService.TAG_FILENAME, photo.id + "." + format)
                intent.putExtra(DownloadService.TAG_ACTION, action)
                startService(intent)
            }
        } else {
            Toast.makeText(this, getString(string.please_wait_for_current_download), Toast.LENGTH_SHORT).show()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            Log.v(TAG, "Permission: " + permissions[0] + "was " + grantResults[0])
            if (action != null) downloadPhoto()
        } else {
            Toast.makeText(this, getString(string.no_permisson), Toast.LENGTH_SHORT).show()
        }
    }

    private fun shareUrl(url: String) {
        val share = Intent(Intent.ACTION_SEND)
        share.type = "text/plain"
        share.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        share.putExtra(Intent.EXTRA_SUBJECT, string.flickr_image)
        share.putExtra(Intent.EXTRA_TEXT, url)
        startActivity(Intent.createChooser(share, getString(string.share_with)))
    }

    private fun hideFAB(fab: FloatingActionButton) {
        val params = fab.layoutParams as LayoutParams
        val behavior = params.behavior as Behavior?
        if (behavior != null) behavior.isAutoHideEnabled = false
        fab.hide()
    }

    private fun showFAB(fab: FloatingActionButton) {
        fab.show()
        val params = fab.layoutParams as LayoutParams
        val behavior = params.behavior as Behavior?
        if (behavior != null) behavior.isAutoHideEnabled = true
    }

    companion object {
        private val TAG = SinglePhotoActivity::class.java.simpleName
        const val PHOTO_TAG = "PARCELABLE_PHOTO"
        const val SAVE_HISTORY_TAG = "SAVE_HISTORY"
        const val PHOTOS_URL = "https://www.flickr.com/photos/"
        private const val PREFERENCE_FIRST_LAUNCH = "SINGLE_PHOTO_FIRST_LAUNCH"
    }
}
